### PR TITLE
[Merged by Bors] - txs: fix missing headers

### LINKED ISSUE #3328: txs: numerous failures due to the missing header
in the new longevity test after applying i am catching numerous failures due to the missing header.

during proposal building:
Screenshot from 2022-07-05 12-02-33

during recovery:
Screenshot from 2022-07-05 12-03-29

### PR DESCRIPTION
**Pull request by dshulyak**
closes: https://github.com/spacemeshos/go-spacemesh/issues/3328

changes in the mempool:
- parse transactions that were previously included into the block but weren't parsable at that time
- skip transactions without headers in get all pending
- error if header not available when block is generated

### CODE REVIEW

File: txs/conservative_state.go [hunk 1]
```diff
@@ -126,22 +126,33 @@ func (cs *ConservativeState) Validation(raw types.RawTx) system.ValidationReques
...
+	if err := cs.tp.Add(tx, received); err != nil {
 		return err
 	}
-	events.ReportNewTx(types.LayerID{}, tx)
-	events.ReportAccountUpdate(tx.Principal)
+	return cs.addToCache(tx, received)
+}
 
-	return cs.cache.Add(tx, received, nil)
+// AddHeader ...
```
Comment (evt) [inline txs/conservative_state.go]: shall we complete the comment may be? :)

**Review by abergasov — APPROVED** [review]

File: sql/transactions/transactions.go [hunk 2]
```diff
@@ -392,7 +371,7 @@ func GetByAddress(db sql.Executor, from, to types.LayerID, address types.Address
 func GetAllPending(db sql.Executor) ([]*types.MeshTransaction, error) {
 	return queryPending(db, `
 		select tx, header, layer, block, timestamp, id from transactions
-		where applied = ?1 order by timestamp asc`,
+		where applied = ?1 and header is not null order by timestamp asc`,
```
Comment (countvonzero) [inline sql/transactions/transactions.go:374]: worth adding a test for this in TestGetAllPending too.

File: txs/txs_builder.go [hunk 3]
```diff
@@ -74,6 +74,9 @@ func extractProposalMetadata(
 				logger.With().Error("failed to find proposal tx", p.LayerIndex, p.ID(), tid, log.Err(err))
 				return nil, fmt.Errorf("get proposal tx: %w", err)
 			}
+			if mtx.TxHeader == nil {
```
Comment (countvonzero) [inline txs/txs_builder.go:77]: please add a test here. this seems easy to rm by fixing merge conflicts.

**Review by countvonzero — APPROVED** [review]

File: txs/conservative_state.go [hunk 4]
```diff
@@ -211,8 +209,8 @@ func (cs *ConservativeState) getTXsToApply(logger log.Log, toApply *types.Block)
 				return nil, fmt.Errorf("applying block %s with invalid tx %s", toApply.ID(), mtx.ID)
 			}
 			mtx.TxHeader = header
-			// updating header also updates principal/nonce indexes
-			if err = transactions.AddHeader(cs.db, mtx.ID, header); err != nil {
+			// add updates header
```
Comment (jonZlotnik) [inline txs/conservative_state.go]: Keeping the case the same provides better comment searchability in jetbrains
```suggestion
			// Add updates header
```

File: sql/transactions/transactions.go [hunk 5]
```diff
@@ -29,7 +29,10 @@ func Add(db sql.Executor, tx *types.Transaction, received time.Time) error {
 	}
 	if _, err = db.Exec(`
 		insert into transactions (id, tx, header, layer, block, principal, nonce, timestamp, applied)
-		values (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)`,
+		values (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9)
+		on conflict(id) do update set 
+		header=?3, principal=?6, nonce=?7 
+		where header is null;`,
```
Comment (jonZlotnik) [inline sql/transactions/transactions.go:35]: Add description in comment about new behavior for the godocs.

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**